pkg/kubectl/rollback.go

```diff
@@ -0,0 +1,136 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// Rollbacker provides an interface for resources that can be rolled back.
+type Rollbacker interface {
+	Rollback(namespace, name string, updatedAnnotations map[string]string, toRevision int64, obj runtime.Object) (string, error)
+}
+
+func RollbackerFor(kind unversioned.GroupKind, c client.Interface) (Rollbacker, error) {
+	switch kind {
+	case extensions.Kind("Deployment"):
+		return &DeploymentRollbacker{c}, nil
+	}
+	return nil, fmt.Errorf("no rollbacker has been implemented for %q", kind)
+}
+
+type DeploymentRollbacker struct {
+	c client.Interface
+}
+
+func (r *DeploymentRollbacker) Rollback(namespace, name string, updatedAnnotations map[string]string, toRevision int64, obj runtime.Object) (string, error) {
+	deploymentRollback := &extensions.DeploymentRollback{
+		Name:               name,
+		UpdatedAnnotations: updatedAnnotations,
+		RollbackTo: extensions.RollbackConfig{
+			Revision: toRevision,
+		},
+	}
+	result := ""
+
+	// Get current events
+	events, err := r.c.Events(namespace).List(api.ListOptions{})
+	if err != nil {
+		return result, err
+	}
+	// Do the rollback
+	if err := r.c.Extensions().Deployments(namespace).Rollback(deploymentRollback); err != nil {
+		return result, err
+	}
+	// Watch for the changes of events
+	watch, err := r.c.Events(namespace).Watch(api.ListOptions{Watch: true, ResourceVersion: events.ResourceVersion})
+	if err != nil {
+		return result, err
+	}
+	result = watchRollbackEvent(watch)
+	return result, err
+}
+
+// containsNewEvent searches events[lastEventCount:] to see if any event's reason is one of the input reasons.
+// If found, return true and the matching event.
+func containsNewEvent(events *api.EventList, lastEventCount int, reasons []string) (bool, *api.Event) {
+	eventCount := len(events.Items)
+	if eventCount <= lastEventCount {
+		return false, nil
+	}
+	for i := eventCount - 1; i >= lastEventCount; i-- {
+		for _, reason := range reasons {
+			if events.Items[i].Reason == reason {
+				return true, &events.Items[i]
+			}
+		}
+	}
+	return false, nil
+}
+
+// watchRollbackEvent watches for rollback events and returns rollback result
+func watchRollbackEvent(w watch.Interface) string {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
```

Comment:
I don't think we should care about closing the channel. It will be garbage-collected anyway once it goes out of scope. Also fyi defer is expensive :) even though we don't really care about it too here